### PR TITLE
SLT-382: Custom nginx redirects

### PR DESCRIPTION
**Pull request by Jancis**
[nginx map directive](http://nginx.org/en/docs/http/ngx_http_map_module.html) is being leveraged for this task. The list of defined redirects is matched with current url and assigned a variable to custom `$redirect_uri` variable in case it does match. Mapping is done outside of `server` scope, so that full url (with protocol and domain name) can be matched too. Then, the `$redirect_uri` variable is used for actual redirection in `server` scope.

There are following test redirects:
```
nginx:
  redirects: |
    /test1 / ;
    /test2 /test2-redirect ;
    http://feature-redirects.drupal-project-k8s.silta.wdr.io/test3 /test3-redirect ;
    '://~feature-redirects.drupal-project-k8s.silta.wdr.io/test4$' /test4-redirect ;
    ~/test5$ /test5-redirect ;
    ~/test6 /test6-redirect ;
    http://feature-redirects-exposedomain.drupal-project-k8s.silta.wdr.io/test7 /test7-redirect ;
```

---

And this is the result for each of them:

 - works, uses absolute path (`/test`, exact matching):
```
➜  ~ curl https://feature-redirects.drupal-project-k8s.silta.wdr.io/test1 -IL
HTTP/1.1 301 Moved Permanently
Location: https://feature-redirects.drupal-project-k8s.silta.wdr.io/
```

 - works, uses absolute path (`/test`, exact matching):
```
➜  ~ curl https://feature-redirects.drupal-project-k8s.silta.wdr.io/test2 -IL 
HTTP/1.1 301 Moved Permanently
Location: https://feature-redirects.drupal-project-k8s.silta.wdr.io/test2-redirect
```

 - works, full url matching
```
➜  ~ curl https://feature-redirects.drupal-project-k8s.silta.wdr.io/test3 -IL
HTTP/1.1 301 Moved Permanently
Location: https://feature-redirects.drupal-project-k8s.silta.wdr.io/test3-redirect
```
 - XX ~works~, full url matching, protocol as wildcard
```
➜  ~ curl https://feature-redirects.drupal-project-k8s.silta.wdr.io/test4 -IL
HTTP/1.1 301 Moved Permanently
Location: https://feature-redirects.drupal-project-k8s.silta.wdr.io/test4-redirect
```

 - Exact path (all exposeDomains)
```
➜  ~ curl https://feature-redirects.drupal-project-k8s.silta.wdr.io/test5 -IL
HTTP/1.1 301 Moved Permanently
Location: https://feature-redirects.drupal-project-k8s.silta.wdr.io/test5-redirect

➜  ~ curl https://feature-redirects.drupal-project-k8s.silta.wdr.io/test5/subpath -IL
HTTP/2 404
```

 - Reacts to all subpaths too:
```
➜  ~ curl https://feature-redirects.drupal-project-k8s.silta.wdr.io/test6 -I
HTTP/1.1 301 Moved Permanently
Location: https://feature-redirects.drupal-project-k8s.silta.wdr.io/test-6-redirect

➜  ~ curl https://feature-redirects.drupal-project-k8s.silta.wdr.io/test6/subpath -I
HTTP/1.1 301 Moved Permanently
Location: https://feature-redirects.drcupal-project-k8s.silta.wdr.io/test-6-redirect
```

 - Only reacts to specific domain
```
➜  ~ curl https://feature-redirects.drupal-project-k8s.silta.wdr.io/test7 -Ik 
HTTP/2 404 

➜  ~ curl https://feature-redirects-exposedomain.drupal-project-k8s.silta.wdr.io/test7 -Ik
HTTP/1.1 301 Moved Permanently
Location: https://feature-redirects-exposedomain.drupal-project-k8s.silta.wdr.io/test7-redirect
```

### CODE REVIEW

File: chart/templates/drupal-configmap.yaml [hunk 1]
```diff
@@ -237,6 +237,8 @@ data:
         add_header                  Strict-Transport-Security max-age=31536000;
         add_header                  X-XSS-Protection '1; mode=block';
 
+        map_hash_bucket_size 128;
```
Comment (floretan) [inline chart/templates/drupal-configmap.yaml]: Why is this needed?

File: silta/silta.yml [hunk 2]
```diff
@@ -9,3 +9,19 @@ varnish:
...
   enabled: true
+
+exposeDomains:
+  - hostname: feature-redirects-exposedomain.drupal-project-k8s.silta.wdr.io
+    ssl:
+      enabled: true
+      issuer: letsencrypt-staging
+
+nginx:
+  redirects: |
```
Comment (floretan) [inline silta/silta.yml]: If we do some preprocessing on these values, wouldn't it make more sense to list them as a yaml array/object so we can cleanly iterate through them?

File: chart/templates/drupal-configmap.yaml (hunk 1 above)
Comment (Jancis) [reply chart/templates/drupal-configmap.yaml]: because nginx didn't start and returned this error: 
> Nginx: could not build map_hash, you should increase map_hash_bucket_size: 64

https://serverfault.com/questions/853102/nginx-could-not-build-map-hash-you-should-increase-map-hash-bucket-size-64
So i set it to the next value or 128 and that helped. 
http://nginx.org/en/docs/http/ngx_http_map_module.html#map_hash_bucket_size

File: silta/silta.yml (hunk 2 above)
Comment (Jancis) [reply silta/silta.yml]: Only in case we actually do preprocessing on them. 

If not, i (personally) would be happy to keep it as a string because some really long boring lists of redirects would need a list item prefix `-` then.

I am not against defining a list too.

File: chart/templates/drupal-configmap.yaml (hunk 1 above)
Comment (Jancis) [reply chart/templates/drupal-configmap.yaml]: If anyone stumbles accross an error, we can extract the `map_hash_bucket_size` as a variable, but until then, I vote for keeping it simple.

**Comment by Jancis** [comment]
Just adding `-` in front of each redirect and making it a list won't work. Processing entries like this is wrong, because it will also find `://` in destination url.
```
{{- if .Values.nginx.redirects }}
    # Custom redirects with full url matching
    map '$scheme://$host$request_uri' $redirect_uri {
        {{- range .Values.nginx.redirects }}
        {{ if contains "://" . }}
        {{ . }}
        {{- end }}
        {{- end }}
    }
    # Custom redirects with full absolute path matching
    map $request_uri $redirect_uri {
        {{- range .Values.nginx.redirects }}
        {{ if not ( contains "://" . ) }}
        {{ . }}
        {{- end }}
        {{- end }}
    }
{{- end }}
```

To counter that, i added `from` and `to` keys, which changes the definition from 
```
nginx:
  redirects: |
    /test1 / ;
    /test2 /test2-redirect ;
    http://feature-redirects.drupal-project-k8s.silta.wdr.io/test3 /test3-redirect ;
    '~feature-redirects.drupal-project-k8s.silta.wdr.io/test4$' /test4-redirect ;
    ~/test5$ /test5-redirect ;
    ~/test6 /test6-redirect ;
    http://feature-redirects-exposedomain.drupal-project-k8s.silta.wdr.io/test7 /test7-redirect ;
```
to
```
nginx:
  redirects:
    - from: /test1 
      to: /
    - from: /test2 
      to: /test2-redirect
    - from: http://feature-redirects.drupal-project-k8s.silta.wdr.io/test3 
      to: /test3-redirect
    - from: '~feature-redirects.drupal-project-k8s.silta.wdr.io/test4$' 
      to: /test4-redirect
    - from: ~/test5$ 
      to: /test5-redirect
    - from: ~/test6 
      to: /test6-redirect
    - from: http://feature-redirects-exposedomain.drupal-project-k8s.silta.wdr.io/test7 
      to: /test7-redirect
```
The later getting too long and too complex for my liking.

**Comment by Jancis** [comment]
@floretan this is open for review again. I changed the structure of redirect definition (see original post, it's edited) and templates accordingly.
When you accept PR, i'll remove values from silta.yml
